src/realm/object-store/impl/collection_notifier.cpp

```diff
@@ -61,32 +63,64 @@ CollectionNotifier::get_modification_checker(TransactionChangeInfo const& info,
     return DeepChangeChecker(info, *root_table, m_related_tables);
 }
 
-void DeepChangeChecker::find_related_tables(std::vector<RelatedTable>& out, Table const& table)
+void DeepChangeChecker::find_related_tables(DeepChangeChecker::RelatedTables& out, Table const& table)
 {
-    auto table_key = table.get_key();
-    if (any_of(begin(out), end(out), [=](auto& tbl) {
-            return tbl.table_key == table_key;
-        }))
-        return;
+    struct ColKeyHasher {
+        std::size_t operator()(const ColKey& key) const
+        {
+            return std::hash<decltype(ColKey::value)>{}(key.value);
+        }
+    };
+    struct LinkInfo {
+        std::unordered_set<ColKey, ColKeyHasher> link_columns;
+        std::unordered_set<TableKey> connected_tables;
+    };
 
-    // We need to add this table to `out` before recurring so that the check
-    // above works, but we can't store a pointer to the thing being populated
-    // because the recursive calls may resize `out`, so instead look it up by
-    // index every time
-    size_t out_index = out.size();
-    out.push_back({table_key, {}});
-
-    for (auto col_key : table.get_column_keys()) {
-        auto type = table.get_column_type(col_key);
-        if (type == type_Link || type == type_LinkList) {
-            out[out_index].links.push_back({col_key.value, type == type_LinkList});
-            find_related_tables(out, *table.get_link_target(col_key));
+    // Build up the complete forward mapping from the back links.
+    // Following forward link columns does not account for TypedLink
+    // values as part of Dictionary<String, Mixed> for example. But
+    // we do not want to assume that all Mixed columns contain links,
+    // so we rely on the fact that if there are any TypedLinks from a
+    // Mixed value, there will be a corresponding backlink column
+    // created at the destination table.
+    std::unordered_map<TableKey, LinkInfo> complete_mapping;
+    Group* group = table.get_parent_group();
+    REALM_ASSERT(group);
+    auto all_table_keys = group->get_table_keys();
+    for (auto it : all_table_keys) {
+        auto it_table = group->get_table(it);
+        REALM_ASSERT(it_table);
+        auto incoming_link_columns = it_table->get_incoming_link_columns();
+        for (auto& incoming_link : incoming_link_columns) {
+            REALM_ASSERT(incoming_link.first);
+            auto& links = complete_mapping[incoming_link.first];
+            links.link_columns.insert(incoming_link.second);
+            links.connected_tables.insert(it_table->get_key());
+        }
+    }
+
+    std::unordered_set<TableKey> tables_to_check = {table.get_key()};
```

Comment:
There's no reason for this to be a set. It's just being used as a really bad stack or queue.